From 0843c6c5f10d7be2cce693c4026440dc5c25c4f4 Mon Sep 17 00:00:00 2001
From: Patrick NC Lin <patrick.nc.lin.wiwynn@gmail.com>
Date: Fri, 14 Mar 2025 14:25:05 +0800
Subject: [PATCH 2/2] Add bmc reboot event log

- Since the BMC currently does not have a reboot event log, adding this
  feature will help with debugging when issues occur.

- Reset the BMC and check the event log - pass

- WatchdogReset
```
root@bmc:~# reboot
Connection to fe80::d604:e6ff:fe34:42c1%eth8 closed by remote host.
Connection to fe80::d604:e6ff:fe34:42c1%eth8 closed.
[patrick@localhost ~]$
[patrick@localhost ~]$ ssh root@fe80::d604:e6ff:fe34:42c1%eth8
root@fe80::d604:e6ff:fe34:42c1%eth8's password:
Last login: Wed Feb  5 03:01:55 2025 from fe80::6d48:3862:ba5e:6a7a%eth0
root@bmc:~#
root@bmc:~#
root@bmc:~# journalctl | grep RebootCause
Feb 05 03:05:23 bmc phosphor-bmc-state-manager[592]: Setting the RebootCause field to xyz.openbmc_project.State.BMC.RebootCause.Watchdog
root@bmc:~# cat /sys/class/watchdog/watchdog0/bootstatus
32

root@bmc:~# mfg-tool log-display | grep -i Reset
"message": "xyz.openbmc_project.State.WatchdogReset",
```

-PowerOnReset
```
root@bmc:~# cat /sys/class/watchdog/watchdog0/bootstatus
0
root@bmc:~# mfg-tool log-display | grep -i Reset
<7> Loading registries from /usr/share/redfish-registry/phosphor-dbus-interfaces
<7> Finding log entries.
<7> Iterating over entries.
<7> Examining xyz.openbmc_project.Logging.Entry at /xyz/openbmc_project/logging/entry/1.
<7> Examining xyz.openbmc_project.Logging.Entry at /xyz/openbmc_project/logging/entry/2.
        "message": "xyz.openbmc_project.State.PowerOnReset",
```

Change-Id: Iea03d771cc06bf632826f07d71b5c9f55460d887
---
 bmc_state_manager.cpp | 30 ++++++++++++++++++++++++------
 1 file changed, 24 insertions(+), 6 deletions(-)

diff --git a/bmc_state_manager.cpp b/bmc_state_manager.cpp
index cfbcb42..ba3137e 100644
--- a/bmc_state_manager.cpp
+++ b/bmc_state_manager.cpp
@@ -275,9 +275,32 @@ BMC::BMCState BMC::currentBMCState(BMCState value)
 
 BMC::RebootCause BMC::lastRebootCause(RebootCause value)
 {
+    std::string errorMsg;
+
     info("Setting the RebootCause field to {LAST_REBOOT_CAUSE}",
          "LAST_REBOOT_CAUSE", value);
 
+    switch (value)
+    {
+        case RebootCause::Software:
+            errorMsg = "xyz.openbmc_project.State.SoftwareReset";
+            break;
+        case RebootCause::Watchdog:
+            errorMsg = "xyz.openbmc_project.State.WatchdogReset";
+            break;
+        case RebootCause::PinholeReset:
+            errorMsg = "xyz.openbmc_project.State.PinholeReset";
+            break;
+        case RebootCause::POR:
+        default:
+            errorMsg = "xyz.openbmc_project.State.PowerOnReset";
+            break;
+    }
+
+    phosphor::state::manager::utils::createError(
+        this->bus, errorMsg,
+        sdbusplus::server::xyz::openbmc_project::logging::Entry::Level::Notice);
+
     return server::BMC::lastRebootCause(value);
 }
 
@@ -350,15 +373,10 @@ void BMC::discoverLastRebootCause()
     // A 0 indicates a pinhole reset occurred
     if (0 == gpioval)
     {
+        // Generate log telling user a pinhole reset has occurred
         info("The BMC reset was caused by a pinhole reset");
         this->lastRebootCause(RebootCause::PinholeReset);
 
-        // Generate log telling user a pinhole reset has occurred
-        const std::string errorMsg = "xyz.openbmc_project.State.PinholeReset";
-        phosphor::state::manager::utils::createError(
-            this->bus, errorMsg,
-            sdbusplus::server::xyz::openbmc_project::logging::Entry::Level::
-                Notice);
         return;
     }
 
-- 
2.52.0

